module/x/gravity/keeper/evidence.go

```diff
@@ -0,0 +1,93 @@
+package keeper
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/cosmos/gravity-bridge/module/x/gravity/types"
+)
+
+func (k Keeper) CheckBadSignatureEvidence(
+	ctx sdk.Context,
+	msg *types.MsgSubmitBadSignatureEvidence) error {
+	var subject types.EthereumSigned
+
+	k.cdc.UnpackAny(msg.Subject, &subject)
+
+	switch subject := subject.(type) {
+	case *types.OutgoingTxBatch:
+		return k.checkBadSignatureEvidenceInternal(ctx, subject, msg.Signature)
+	case *types.Valset:
+		return k.checkBadSignatureEvidenceInternal(ctx, subject, msg.Signature)
+	case *types.OutgoingLogicCall:
+		return k.checkBadSignatureEvidenceInternal(ctx, subject, msg.Signature)
+
+	default:
+		return sdkerrors.Wrap(types.ErrInvalid, "Bad signature must be over a batch, valset, or logic call")
+	}
+}
+
+func (k Keeper) checkBadSignatureEvidenceInternal(ctx sdk.Context, subject types.EthereumSigned, signature string) error {
+	// Get checkpoint of the supposed bad signature (fake valset, batch, or logic call submitted to eth)
+	gravityID := k.GetGravityID(ctx)
+	checkpoint := subject.GetCheckpoint(gravityID)
+
+	// Try to find the checkpoint in the archives. If it exists, we don't slash because
+	// this is not a bad signature
+	if k.GetPastEthSignatureCheckpoint(ctx, checkpoint) {
```

Comment:
ditto, this will only fetch the latest stored checkpoint

Reply:
See my comment about sets below